apps/privacy-panel/js/app_list.js

```diff
@@ -0,0 +1,327 @@
+
+  var _ = navigator.mozL10n.get;
+  var _lang = navigator.mozL10n.language.code;
+
+
+  /**
+   * Supported application sorting methods:
+   * by name, by trust level, by developer name.
+   * Note: sorting by name works for permissions as well.
+   */
+  var _orderBy = {
+    name: (a, b) => a.name.localeCompare(b.name, _lang), // default
+    trust: (a, b) => a.trust > b.trust, // 'certified', 'privileged', 'web'
+    vendor: (a, b) => a.vendor.localeCompare(b.vendor, _lang)
+  };
+
+
+  /****************************************************************************
+   * Applications - private helpers
+   */
+
+  var _applications = []; // array of {DOMApplication} instances
+  var _defaultIconURL = '../style/images/default.png';
+
+  /**
+   * Get the list of installed apps.
+   */
+  function _getApplications(onsuccess, onerror) {
+    onsuccess = typeof onsuccess === 'function' ? onsuccess : function() {};
+    onerror = typeof onerror === 'function' ? onerror : function() {};
+
+    var mozAppsMgmt = navigator.mozApps && navigator.mozApps.mgmt;
+    if (!mozAppsMgmt) {
+      console.error('navigator.mozApps.mgmt is undefined');
+      onerror();
+      return;
+    }
+
+    var req = mozAppsMgmt.getAll();
+    req.onerror = onerror;
+    req.onsuccess = event => {
+      _applications = event.target.result.map(_defCustomProperties)
+                                         .sort(_orderBy.name);
+      onsuccess();
+    };
+  }
+
+  /**
+   * Get the app icon that best suits the device display size.
+   */
+  function _getBestIconURL(app) {
+    var icons = (app.manifest || app.updateManifest).icons;
+    if (!icons || !Object.keys(icons).length) {
+      return _defaultIconURL;
+    }
+
+    // The preferred size is 30 pixels by default.
+    // On an HDPI device, we may use a larger size than 30 * 1.5 = 45 pixels.
+    var preferredIconSize = 30 * (window.devicePixelRatio || 1);
+    var preferredSize = Number.MAX_VALUE;
+    var max = 0;
+
+    for (var size in icons) {
+      size = parseInt(size, 10);
+      if (size > max) {
+        max = size;
+      }
+      if (size >= preferredIconSize && size < preferredSize) {
+        preferredSize = size;
+      }
+    }
+
+    // If there is an icon matching the preferred size, we return the result,
+    // if there isn't, we will return the maximum available size.
+    if (preferredSize === Number.MAX_VALUE) {
+      preferredSize = max;
+    }
+
+    var url = icons[preferredSize];
+    if (!url) {
+      return _defaultIconURL;
+    }
+    return !(/^(http|https|data):/.test(url)) ? app.origin + url : url;
+  }
+
+  /**
+   * Add properties to the {DOMApplication} instances in the _application list.
+   *   .name:        localized name
+   *   .trust:       trust level (= certified, privileged, web)
+   *   .vendor:      developer name
+   *   .iconURL:     URL of the best icon for the current display
+   *   .permissions: filtered list of permissions that are actually used
+   */
```

Comment:
How about making this a `_makeAppRepresentation` returning new objects with just the properties we need.
Or, alternatively, a clone of non-functions properties + our custom ones.

(yes, what prompted this bikeshed is the fact that this is called in the middle of a very functional chain but is then having `defineProperty`-level side effect ;))